File: app/lib/models/transaction.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'transaction.freezed.dart';
part 'transaction.g.dart';

enum TransactionType {
  @JsonValue('expense')
  expense,
  @JsonValue('income')
  income,
}

@freezed
class Transaction with _$Transaction {
  const factory Transaction({
    required String id,
    required TransactionType type,
    required double amount,
    required String currencyCode,
    required String category,
    @JsonKey(name: 'counterparty', readValue: _readCounterparty)
    String? counterparty,
    required DateTime date,
    @JsonKey(name: 'placeName', readValue: _readPlaceName)
    String? location,
    @Default(0) int regretLevel,
    String? notes,
    required DateTime createdAt,
    double? exchangeRateToBase,
  }) = _Transaction;

  factory Transaction.fromJson(Map<String, dynamic> json) =>
      _$TransactionFromJson(json);
}

Object? _readCounterparty(Map json, String _) =>
    json['counterparty'] ?? json['merchant'] ?? json['description'];

Object? _readPlaceName(Map json, String _) =>
    json['placeName'] ?? json['merchantName'] ?? json['location'];

File: src/Conscia.Domain/Entities/Transaction.cs
using Conscia.Domain.Enums;
using Conscia.Domain.ValueObjects;

namespace Conscia.Domain.Entities;

public class Transaction
{
    public Guid Id { get; set; }
    public Guid UserId { get; set; }
    public TransactionType Type { get; set; }
    public Money Amount { get; set; } = null!;
    public string Category { get; set; } = string.Empty;
    public string? Counterparty { get; set; }
    public DateTime Date { get; set; }
    public Location? Location { get; set; }
    public RegretLevel? RegretLevel { get; set; }
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
}

public class Location
{
    public double Latitude { get; set; }
    public double Longitude { get; set; }
    public string? PlaceName { get; set; }
}
